File: kotlin/src/main/kotlin/benchmarks/Mandelbrot.kt
package benchmarks

import Benchmark

class Mandelbrot : Benchmark() {
    private var w: Int = 0
    private var h: Int = 0
    private lateinit var result: java.io.ByteArrayOutputStream

    companion object {
        private const val ITER = 50
        private const val LIMIT = 2.0
    }

    init {
        w = configVal("w").toInt()
        h = configVal("h").toInt()
    }

    override fun prepare() {
        result = java.io.ByteArrayOutputStream()
    }

    override fun run(iterationId: Int) {        

        result.write("P4\n$w $h\n".toByteArray())

        var bitNum = 0
        var byteAcc = 0

        for (y in 0 until h) {
            for (x in 0 until w) {
                var zr = 0.0
                var zi = 0.0
                var tr = 0.0
                var ti = 0.0
                val cr = 2.0 * x / w.toDouble() - 1.5
                val ci = 2.0 * y / h.toDouble() - 1.0

                var i = 0
                while (i < ITER && tr + ti <= LIMIT * LIMIT) {
                    zi = 2.0 * zr * zi + ci
                    zr = tr - ti + cr
                    tr = zr * zr
                    ti = zi * zi
                    i += 1
                }

                byteAcc = (byteAcc shl 1) and 0xFF
                if (tr + ti <= LIMIT * LIMIT) {
                    byteAcc = byteAcc or 0x01
                }
                bitNum += 1

                if (bitNum == 8) {
                    result.write(byteAcc)
                    byteAcc = 0
                    bitNum = 0
                } else if (x == w - 1) {

                    if (bitNum > 0) {
                        byteAcc = (byteAcc shl (8 - bitNum)) and 0xFF
                        result.write(byteAcc)
                    }
                    byteAcc = 0
                    bitNum = 0
                }
            }
        }
    }

    override fun checksum(): UInt {
        val bytes = result.toByteArray()
        val checksum = Helper.checksum(bytes)        
        return checksum
    }

    override fun name(): String = "Mandelbrot"
}

File: rust/src/benchmarks/mandelbrot.rs
use super::super::{helper, Benchmark};
use crate::config_i64;

const ITER: i32 = 50;
const LIMIT: f64 = 2.0;

pub struct Mandelbrot {
    w: i64,
    h: i64,
    result_bin: Vec<u8>,
}

impl Mandelbrot {
    pub fn new() -> Self {
        let w = config_i64("CLBG::Mandelbrot", "w");
        let h = config_i64("CLBG::Mandelbrot", "h");

        Self {
            w,
            h,
            result_bin: Vec::new(),
        }
    }
}

impl Benchmark for Mandelbrot {
    fn name(&self) -> String {
        "CLBG::Mandelbrot".to_string()
    }

    fn run(&mut self, _iteration_id: i64) {
        let w = self.w as usize;
        let h = self.h as usize;

        let header = format!("P4\n{} {}\n", w, h);
        self.result_bin.extend_from_slice(header.as_bytes());

        let mut bit_num = 0;
        let mut byte_acc: u8 = 0;

        for y in 0..h {
            for x in 0..w {
                let mut zr = 0.0;
                let mut zi = 0.0;
                let mut tr = 0.0;
                let mut ti = 0.0;

                let cr = (2.0 * x as f64 / w as f64) - 1.5;
                let ci = (2.0 * y as f64 / h as f64) - 1.0;

                let mut i = 0;
                while i < ITER && (tr + ti) <= (LIMIT * LIMIT) {
                    zi = 2.0 * zr * zi + ci;
                    zr = tr - ti + cr;
                    tr = zr * zr;
                    ti = zi * zi;
                    i += 1;
                }

                byte_acc <<= 1;
                if tr + ti <= LIMIT * LIMIT {
                    byte_acc |= 0x01;
                }
                bit_num += 1;

                if bit_num == 8 {
                    self.result_bin.push(byte_acc);
                    byte_acc = 0;
                    bit_num = 0;
                } else if x == w - 1 {
                    byte_acc <<= 8 - (w % 8);
                    self.result_bin.push(byte_acc);
                    byte_acc = 0;
                    bit_num = 0;
                }
            }
        }
    }

    fn checksum(&self) -> u32 {
        helper::checksum_bytes(&self.result_bin)
    }
}
